From 6da0398e9cacf73d23fc89800570bbec1a221a9f Mon Sep 17 00:00:00 2001
From: Michal Srb <michalsrb@gmail.com>
Date: Tue, 21 Feb 2017 23:12:42 +0200
Subject: [PATCH 06/13] Add FcPreprocessString speeding up FcCompareString.

FcPreprocessString calculates hash of the string to fast reject not matching strings.

TODO: Add benchmark numbers.
---
 src/fcmatch.c | 17 +++++++++++++++++
 src/fcobjs.h  |  8 ++++----
 2 files changed, 21 insertions(+), 4 deletions(-)

diff --git a/src/fcmatch.c b/src/fcmatch.c
index 0a4ca86..474aaaf 100644
--- a/src/fcmatch.c
+++ b/src/fcmatch.c
@@ -55,9 +55,26 @@ FcCompareNumber (FcValue *value1, FcPrepValue *prep1, FcValue *value2, FcPrepVal
     return v;
 }
 
+static FcPrepValue
+FcPreprocessString (FcValue *v)
+{
+    FcPrepValue prep;
+    prep.type = FcPrepStrHashIgnoreCase;
+    prep.str_hash = FcStrHashIgnoreCase(FcValueString(v));
+    return prep;
+}
+
 static double
 FcCompareString (FcValue *v1, FcPrepValue *p1, FcValue *v2, FcPrepValue *p2)
 {
+    if (p1->type == FcPrepStrHashIgnoreCase &&
+	p2->type == FcPrepStrHashIgnoreCase)
+    {
+	// If hashes are not matching, return fast
+	if (p1->str_hash != p2->str_hash)
+	    return 1.0;
+    }
+
     return (double) FcStrCmpIgnoreCase (FcValueString(v1), FcValueString(v2)) != 0;
 }
 
diff --git a/src/fcobjs.h b/src/fcobjs.h
index cf93c07..5b16c2b 100644
--- a/src/fcobjs.h
+++ b/src/fcobjs.h
@@ -24,7 +24,7 @@
 /* DON'T REORDER!  The order is part of the cache signature. */
 FC_OBJECT (FAMILY,		FcTypeString,	FcCompareFamily,	NULL)
 FC_OBJECT (FAMILYLANG,		FcTypeString,	NULL,			NULL)
-FC_OBJECT (STYLE,		FcTypeString,	FcCompareString,	NULL)
+FC_OBJECT (STYLE,		FcTypeString,	FcCompareString,	FcPreprocessString)
 FC_OBJECT (STYLELANG,		FcTypeString,	NULL,			NULL)
 FC_OBJECT (FULLNAME,		FcTypeString,	NULL,			NULL)
 FC_OBJECT (FULLNAMELANG,	FcTypeString,	NULL,			NULL)
@@ -35,7 +35,7 @@ FC_OBJECT (SIZE,		FcTypeRange,	FcCompareSizeRange,	NULL)
 FC_OBJECT (ASPECT,		FcTypeDouble,	NULL,			NULL)
 FC_OBJECT (PIXEL_SIZE,		FcTypeDouble,	FcCompareSize,		NULL)
 FC_OBJECT (SPACING,		FcTypeInteger,	FcCompareNumber,	NULL)
-FC_OBJECT (FOUNDRY,		FcTypeString,	FcCompareString,	NULL)
+FC_OBJECT (FOUNDRY,		FcTypeString,	FcCompareString,	FcPreprocessString)
 FC_OBJECT (ANTIALIAS,		FcTypeBool,	FcCompareBool,		NULL)
 FC_OBJECT (HINT_STYLE,		FcTypeInteger,	NULL,			NULL)
 FC_OBJECT (HINTING,		FcTypeBool,	NULL,			NULL)
@@ -44,7 +44,7 @@ FC_OBJECT (AUTOHINT,		FcTypeBool,	NULL,			NULL)
 FC_OBJECT (GLOBAL_ADVANCE,	FcTypeBool,	NULL,			NULL)	/* deprecated */
 FC_OBJECT (FILE,		FcTypeString,	FcCompareFilename,	NULL)
 FC_OBJECT (INDEX,		FcTypeInteger,	NULL,			NULL)
-FC_OBJECT (RASTERIZER,		FcTypeString,	FcCompareString,	NULL)	/* deprecated */
+FC_OBJECT (RASTERIZER,		FcTypeString,	FcCompareString,	FcPreprocessString)	/* deprecated */
 FC_OBJECT (OUTLINE,		FcTypeBool,	FcCompareBool,		NULL)
 FC_OBJECT (SCALABLE,		FcTypeBool,	FcCompareBool,		NULL)
 FC_OBJECT (DPI,			FcTypeDouble,	NULL,			NULL)
@@ -58,7 +58,7 @@ FC_OBJECT (CHARSET,		FcTypeCharSet,	FcCompareCharSet,	NULL)
 FC_OBJECT (LANG,		FcTypeLangSet,	FcCompareLang,		NULL)
 FC_OBJECT (FONTVERSION,		FcTypeInteger,	FcCompareNumber,	NULL)
 FC_OBJECT (CAPABILITY,		FcTypeString,	NULL,			NULL)
-FC_OBJECT (FONTFORMAT,		FcTypeString,	FcCompareString,	NULL)
+FC_OBJECT (FONTFORMAT,		FcTypeString,	FcCompareString,	FcPreprocessString)
 FC_OBJECT (EMBOLDEN,		FcTypeBool,	NULL,			NULL)
 FC_OBJECT (EMBEDDED_BITMAP,	FcTypeBool,	NULL,			NULL)
 FC_OBJECT (DECORATIVE,		FcTypeBool,	FcCompareBool,		NULL)
-- 
2.15.1

